### PR TITLE
Refactor for ease of safe use (backport)

### PR DESCRIPTION
**Pull request by thomassa**
Backported from master:
+ Add functions for ease of safe use
+ Refactor to use them.

### CODE REVIEW

**Comment by thomassa** [comment]
Note that the first six commits are cherry-picks from commits already in https://github.com/xapi-project/nbd/commits/master
but the next five (the ones for `with_block`, `with_connection` and `with_channel` plus the final fix) are not yet: this pull-request is the first review for those.

**Comment by mseri** [comment]
Wouldn't it be better to update fully the upstream version and then, backport what is needed instead of diverging so much?

**Comment by thomassa** [comment]
The first consumer of this new functionality (apart from the CLI tool inside this repo) is xapi-nbd, so these extensions to the interface should be judged in the context of xapi-nbd's use of them.

Sadly, xapi-nbd is limited to using the v2-based branch at present (due to OCaml versions and perhaps other constraints).

**Comment by mseri** [comment]
We have merged the changes for the new compiler, so this may be fixable soon

**Comment by gaborigloi** [comment]
Typo in comit: `Add `Server.with_connection` function`:
This calls `Server.connect`

File: lib/server.ml [hunk 1]
```diff
@@ -139,6 +139,15 @@ let connect channel ?offer () =
         else negotiate_tls make_tls_channel
       )
 
+let with_connection clearchan ?offer f =
+  (match offer with
+    | None -> connect clearchan ()
+    | Some offer -> connect clearchan ~offer ())
+  >>= fun (_name, t) ->
```
Comment (gaborigloi) [inline lib/server.ml]: We need to pass `_name` to `f`, because `xapi-nbd` will need to know the export name.

File: lib/server.ml [hunk 2]
```diff
@@ -139,6 +139,15 @@ let connect channel ?offer () =
         else negotiate_tls make_tls_channel
       )
 
+let with_connection clearchan ?offer f =
+  (match offer with
+    | None -> connect clearchan ()
+    | Some offer -> connect clearchan ~offer ())
```
Comment (gaborigloi) [inline lib/server.ml]: Luckily we can simplify propagating optional parameters with:
`connect clearchan ?offer ()` :smile:

**Comment by thomassa** [comment]
In any case, for now I shall make a smaller pull-request that backports only the commits that are already in `master`.

File: lwt/nbd_lwt_unix.mli [hunk 3]
```diff
@@ -35,6 +35,15 @@ val with_block: string -> (Block.t -> 'a Block.io) -> 'a Block.io
 (** [with_block filename f] calls [Block.connect filename] and applies [f] to the result,
     with a guarantee to call [Block.disconnect] afterwards. *)
 
+val with_channel:
+  Lwt_unix.file_descr ->
+  tls_role option ->
+  (Nbd.Channel.cleartext_channel -> 'a Block.io) ->
+  'a Block.io
+(** [with_fd fd role f] calls [cleartext_channel_of_fd fd role] then
```
Comment (gaborigloi) [inline lwt/nbd_lwt_unix.mli]: typo: `[with_channel ...`

File: lwt/nbd_lwt_unix.ml [hunk 4]
```diff
@@ -112,10 +112,13 @@ let init_tls_get_ctx ~certfile ~ciphersuites =
 
 let with_block filename f =
   Block.connect filename
-  >>= fun b ->
-  Lwt.finalize
-    (fun () -> f b)
-    (fun () -> Block.disconnect b)
+  >>= function
+  | `Error _ ->
+    Lwt.fail_with (Printf.sprintf "with_block failed in to open %s" filename)
```
Comment (gaborigloi) [inline lwt/nbd_lwt_unix.ml]: Typo: "failed in to"

**Comment by thomassa** [comment]
Let's get #81 merged first, shall we?

File: lib/server.ml (hunk 2 above)
Comment (thomassa) [reply lib/server.ml]: Oops, yes, that's why the function has that last parameter.

**Comment by thomassa** [comment]
This is a pure backport from master now.
It still builds, and the code in https://github.com/xapi-project/xapi-nbd/pull/13 still builds on top of it.

**Comment by gaborigloi** [comment]
Missing commit: "Use the new with_block function"

**Comment by gaborigloi** [comment]
There is some more code that is common between xapi-nbd and nbd: we could factor out the main loop accepting clients:
```
 let t =
       let sock = Lwt_unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
       Lwt.finalize
         (fun () ->
            Lwt_unix.setsockopt sock Lwt_unix.SO_REUSEADDR true;
            let sockaddr = Lwt_unix.ADDR_INET(Unix.inet_addr_any, port) in
            Lwt_unix.Versioned.bind_2 sock sockaddr >>= fun () ->
            Lwt_unix.listen sock 5;
            let rec loop () =
              Lwt_unix.accept sock
              >>= fun (fd, exportname) ->
              (* Background thread per connection *)
              let _ = 
                Lwt.catch
                  (fun () -> handle_connection fd exportname) 
                  (fun e -> Lwt_io.eprintf "Caught exception %s while handling connection\n%!" (Printexc.to_string e)) 
              in  
              loop ()
            in  
            loop ()
         )   
         (ignore_exn (fun () -> Lwt_unix.close sock))
     in  
     Lwt_main.run t
```
And `handle_connection` can call `serve` if it decides to accept the given `exportname`.

**Comment by thomassa** [comment]
I've added a new commit to use `with_block` but not yet built-and-tested...

**Comment by samoht** [comment]
just out of curiosity: how do you generate these reports while using jbuilder? Would be useful to share your recipe as it seems that the community is a bit stuck with it currently (see https://github.com/janestreet/jbuilder/issues/57)

**Comment by mseri** [comment]
I did share the recipe with somebody. Unfortunately is kind of a hack: we autogenerate the ppx flags for old ppx packages and autogenerate the bisect pps stanza in the jbuild file when `BISECT_ENABLE` is 'YES`: see https://github.com/xapi-project/nbd/blob/bugfix-v2.x/lib/jbuild

The `bisect_ppx` package is not declared in the opam file and is installed by hand when running the coverage builds (as part of a travis post install hook): see https://github.com/xapi-project/nbd/blob/master/.coverage.sh

Not nice but it does the job until we get a final solution.

**Comment by samoht** [comment]
I see thanks. Would be indeed great to have a better solution at one point!

**Pull request merged**